src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsServerRouter/DiagnosticsServerRouter.cs

```diff
@@ -0,0 +1,841 @@
+    internal class RuntimeConnectTimeoutException : TimeoutException
+    {
+        public RuntimeConnectTimeoutException(int TimeoutMs)
+            : base(string.Format("No new runtime endpoints connected, waited {0} ms", TimeoutMs))
+        { }
+    }
+
+    internal class BackendStreamConnectTimeoutException : TimeoutException
+    {
+        public BackendStreamConnectTimeoutException(int TimeoutMs)
+            : base(string.Format("No new backend streams available, waited {0} ms", TimeoutMs))
+        { }
+    }
+
+    /// <summary>
+    /// Base class representing a Diagnostics Server router.
+    /// </summary>
+    internal class DiagnosticsServerRouter
+    {
+        protected readonly ILogger _logger;
+
+        public DiagnosticsServerRouter(ILogger logger)
+        {
+            _logger = logger;
+        }
+
+        public ILogger Logger
+        {
+            get { return _logger; }
+        }
+
+        public virtual void Start()
+        {
+            throw new NotImplementedException();
+        }
+
+        public virtual Task Stop()
+        {
+            throw new NotImplementedException();
+        }
+
+        public virtual void Reset()
+        {
+            throw new NotImplementedException();
+        }
+
+        public virtual Task<ConnectedRouter> ConnectRouterAsync(CancellationToken token)
+        {
+            throw new NotImplementedException();
+        }
+    }
+
+    /// <summary>
+    /// This class represent a TCP/IP server endpoint used when building up router instances.
+    /// </summary>
+    internal class TcpServerRouter : DiagnosticsServerRouter
+    {
+        protected string _tcpServerAddress;
+
+        protected ReversedDiagnosticsServer _tcpServer;
+        protected IpcEndpointInfo _tcpServerEndpointInfo;
+
+        public int RuntimeInstanceConnectTimeoutMs { get; set; } = 60000;
+        public int TcpServerConnectTimeoutMs { get; set; } = 5000;
+
+        public Guid RuntimeInstanceId
+        {
+            get { return _tcpServerEndpointInfo.RuntimeInstanceCookie; }
+        }
+
+        public int RuntimeProcessId
+        {
+            get { return _tcpServerEndpointInfo.ProcessId; }
+        }
+
+        public string TcpServerAddress
+        {
+            get { return _tcpServerAddress; }
+        }
+
+        protected TcpServerRouter(string tcpServer, int runtimeTimeoutMs, ILogger logger)
+            : base(logger)
+        {
+            _tcpServerAddress = tcpServer;
+            if (string.IsNullOrEmpty(_tcpServerAddress))
+            {
+                _tcpServerAddress = "127.0.0.1:0";
+            }
+            else
+            {
+                _tcpServerAddress = _tcpServerAddress.Replace("tcp://", "");
+                _tcpServerAddress = _tcpServerAddress.Replace(Uri.UriSchemeNetTcp, "");
+            }
+
+            RuntimeInstanceConnectTimeoutMs = runtimeTimeoutMs;
+
+            _tcpServer = new ReversedDiagnosticsServer(_tcpServerAddress, true);
```

Comment:
nit: 
```suggestion
            _tcpServer = new ReversedDiagnosticsServer(_tcpServerAddress, enableTcpIpProtocol : true);
```

Reply:
For this case I can agree to specify parameter name since it has some specific meaning, so highlighting that make sense.